### ISSUE TITLE
explicit_counter_loop false positive with continue statement

### ISSUE THREAD

**Issue by mbrt** [rust-lang-nursery/rust-clippy#1219]
Clippy shows a false positive in this example, asking for an `explicit_counter_loop` refactor:

```
    let text = "banana";
    let mut count = 0;
    for ch in text.chars() {
        if ch == 'a' {
            continue;
        }
        count += 1
    }
    println!("{}", count);
```

But since the `continue` statement skips the counter, the proposed `for (count, ch) in text.char_indices()` is not equivalent.

**Comment by mcarton** [comment]
We could suggest to use `filter` when the first statement in a `for` loop is `if f(var) { continue }`.

**Comment by mbrt** [comment]
Actually I think that you can simply suppress the suggestion if a continue
statement is present in the loop body. There could be dependencies (in term
of side effects) among the counter and continue and break statements. It's
not easy to detect when they are effectively equivalent.

Il 08 set 2016 23:53, "Martin Carton" <notifications@github.com> ha scritto:

**Comment by Reylak** [comment]
I agree with @mbrt, I think this lint in most often wrong whenever the counter update statement can be unreachable because of a `continue`.

**Comment by oli-obk** [comment]
Yea let's just bail on `continue` statements
